phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -495,7 +524,13 @@ public Boolean call() throws Exception {
                         exceptionMessage = "Pool closed, not attempting to verify index rows! " + indexHTable.getName();
                         throw new IOException(exceptionMessage);
                     }
-                    verifyIndexRows(keys);
+                    verifyIndexRows(keys, perTaskRowKeyToDataPutMap);
+                    if (verifyType == IndexTool.IndexVerifyType.BEFORE || verifyType == IndexTool.IndexVerifyType.BOTH) {
+                        synchronized (dataKeyToDataPutMap) {
```

Comment:
There is no concurrent map for tree maps (i.e., sorted maps). Please note that this is called after verifying all rows per task. So, it should not lead to a lock contention issue as it will not be a frequent operation.